packages/client/wallets/smart-wallet/src/blockchain/wallets/SendTransactionService.ts

```diff
@@ -0,0 +1,191 @@
+     */
+    confirmations: number;
+    /**
+     * The timeout in milliseconds to wait for a transaction to confirm before throwing an error.
+     */
+    transactionConfirmationTimeout: number;
+}
+
+export class SendTransactionService {
+    constructor(
+        private publicClient: PublicClient,
+        private defaultSendTransactionOptions: SendTransactionOptions = {
+            confirmations: 2,
+            transactionConfirmationTimeout: 30_000,
+        }
+    ) {}
+
+    async sendTransaction(
+        request: TransactionServiceTransactionRequest,
+        client: SmartAccountClient<EntryPoint, Transport, Chain, SmartAccount<EntryPoint>>,
+        config: Partial<SendTransactionOptions> = {}
+    ): Promise<Hex> {
+        const { confirmations, transactionConfirmationTimeout } = this.getConfig(config);
+
+        // Simulate
+        await this.simulateCall(request, undefined, "simulation");
+
+        // Send
+        let hash;
+        try {
+            hash = await client.writeContract({
+                ...request,
+                account: client.account,
+                chain: client.chain,
+            });
+        } catch (e) {
+            if (e instanceof BaseError) {
+                throw new EVMSendTransactionError(e.message, e, "send");
+            }
+            throw e;
+        }
+
+        // Confirm
+        try {
+            const receipt = await this.publicClient.waitForTransactionReceipt({
+                hash,
+                confirmations,
+                timeout: transactionConfirmationTimeout,
+            });
+            return await this.handleReceipt(receipt, request);
+        } catch (e) {
+            if (e instanceof BaseError) {
+                throw new EVMSendTransactionError(e.message, e, "confirmation");
+            }
+            throw e;
+        }
+    }
+
+    private getConfig(config: Partial<SendTransactionOptions>): SendTransactionOptions {
+        return {
+            ...this.defaultSendTransactionOptions,
+            ...config,
+        };
+    }
+
+    private async handleReceipt(receipt: TransactionReceipt, request: TransactionServiceTransactionRequest) {
+        if (receipt.status === "reverted") {
+            // This should revert and throw the full reason
+            await this.simulateCall(request, receipt.transactionHash, "execution");
+            // Otherwise, throw a generic error (this should practically never happen)
+            throw new EVMSendTransactionExecutionRevertedError(
+                "Transaction reverted but unable to detect the reason",
+                new ContractFunctionRevertedError({ abi: request.abi as Abi, functionName: request.functionName }),
+                new ContractFunctionRevertedError({ abi: request.abi as Abi, functionName: request.functionName }),
+                receipt.transactionHash,
+                "execution"
+            );
+        }
+        return receipt.transactionHash;
```

Comment:
Would be good to test this locally if you haven't already

Reply:
It's practically impossible to do so under AA since the bundler etc simulate the transaction before it lands, but I tested all of the components that get called here (e.g. that waitForTransaction doesn't throw an error on revert and returns the status "reverted") and unit tested that they work together in this flow

Reply:
ok!

Reply:
If it's practically impossible to exercise this code, is it worth having?

Reply:
I think it's good to have a branch that handles this case so we never yield a tx hash that contains a reverted transaction, even if we can't simulate it ourselves under ideal conditions